src/pem.rs

```diff
@@ -3,65 +3,178 @@ use alloc::format;
 use alloc::string::String;
 use alloc::vec;
 use alloc::vec::Vec;
-#[cfg(feature = "std")]
-use core::iter;
+use core::marker::PhantomData;
 use core::ops::ControlFlow;
 #[cfg(feature = "std")]
 use std::io::{self, ErrorKind};
 
 use crate::base64;
-use crate::{
-    CertificateDer, CertificateRevocationListDer, CertificateSigningRequestDer, PrivatePkcs1KeyDer,
-    PrivatePkcs8KeyDer, PrivateSec1KeyDer, SubjectPublicKeyInfoDer,
-};
 
-/// The contents of a single recognised block in a PEM file.
-#[non_exhaustive]
-#[derive(Debug, PartialEq)]
-pub enum Item {
-    /// A DER-encoded x509 certificate.
+/// Items that can be decoded from PEM data.
+pub trait PemObject: Sized {
+    /// Decode the first section of this type from PEM contained in
+    /// a byte slice.
     ///
-    /// Appears as "CERTIFICATE" in PEM files.
-    X509Certificate(CertificateDer<'static>),
+    /// `Err(NoItemsFound)` is returned if no such items are found.
+    fn from_pem_slice(pem: &[u8]) -> Result<Self, Error> {
+        Self::pem_slice_iter(pem)
+            .next()
+            .unwrap_or(Err(Error::NoItemsFound))
+    }
 
-    /// A DER-encoded Subject Public Key Info; as specified in RFC 7468.
-    ///
-    /// Appears as "PUBLIC KEY" in PEM files.
-    SubjectPublicKeyInfo(SubjectPublicKeyInfoDer<'static>),
+    /// Iterate over all sections of this type from PEM contained in
+    /// a byte slice.
+    fn pem_slice_iter(pem: &[u8]) -> SliceIter<'_, Self> {
+        SliceIter {
+            current: pem,
+            _ty: PhantomData,
+        }
+    }
 
-    /// A DER-encoded plaintext RSA private key; as specified in PKCS #1/RFC 3447
+    /// Decode the first section of this type from the PEM contents of the named file.
     ///
-    /// Appears as "RSA PRIVATE KEY" in PEM files.
-    Pkcs1Key(PrivatePkcs1KeyDer<'static>),
+    /// `Err(NoItemsFound)` is returned if no such items are found.
+    #[cfg(feature = "std")]
+    fn from_pem_file(file_name: impl AsRef<std::path::Path>) -> Result<Self, Error> {
+        Self::pem_file_iter(file_name)?
+            .next()
+            .unwrap_or(Err(Error::NoItemsFound))
+    }
 
-    /// A DER-encoded plaintext private key; as specified in PKCS #8/RFC 5958
+    /// Iterate over all sections of this type from the PEM contents of the named file.
     ///
-    /// Appears as "PRIVATE KEY" in PEM files.
-    Pkcs8Key(PrivatePkcs8KeyDer<'static>),
+    /// Unlike other `iter_*` functions, this eagerly consumes all the sections
+    /// to bring together errors opening the file and parsing the contents.
+    #[cfg(feature = "std")]
+    fn pem_file_iter(
+        file_name: impl AsRef<std::path::Path>,
+    ) -> Result<ReadIter<io::BufReader<std::fs::File>, Self>, Error> {
+        Ok(ReadIter::<_, Self> {
+            rd: io::BufReader::new(std::fs::File::open(file_name).map_err(Error::Io)?),
```

Comment:
Nit: we've (conditionally) imported `io`, but are locally qualifying `File`. I think I have a slight preference for also importing `File`.